Submitted By:            Bruce Dubbs <bdubbs_at_linuxfromscratch_dot_org>
Updated By:              Armin K. <krejzi at email dot com>
Date:                    2014-10-11
Initial Package Version: 4.3
Upstream Status:         Already in upstream patch repo
Origin:                  Upstream
Description:             This patch contains upstream patch numbers 001 thru 030
                         and fixes for CVE-2014-6271, CVE-2014-6277, CVE-2014-6278,
                         CVE-2014-7169, and CVE-2014-7187

--- a/arrayfunc.c	2013-08-02 22:19:59.000000000 +0200
+++ b/arrayfunc.c	2014-10-11 10:34:38.173610936 +0200
@@ -179,6 +179,7 @@
     array_insert (array_cell (entry), ind, newval);
   FREE (newval);
 
+  VUNSETATTR (entry, att_invisible);	/* no longer invisible */
   return (entry);
 }
 
@@ -597,6 +598,11 @@
       if (assoc_p (var))
 	{
 	  val = expand_assignment_string_to_string (val, 0);
+	  if (val == 0)
+	    {
+	      val = (char *)xmalloc (1);
+	      val[0] = '\0';	/* like do_assignment_internal */
+	    }
 	  free_val = 1;
 	}
 
--- a/bashline.c	2014-02-10 01:56:58.000000000 +0100
+++ b/bashline.c	2014-10-11 10:34:38.173610936 +0200
@@ -4167,9 +4167,16 @@
   int qc;
 
   qc = rl_dispatching ? rl_completion_quote_character : 0;  
-  dfn = bash_dequote_filename ((char *)text, qc);
+  /* If rl_completion_found_quote != 0, rl_completion_matches will call the
+     filename dequoting function, causing the directory name to be dequoted
+     twice. */
+  if (rl_dispatching && rl_completion_found_quote == 0)
+    dfn = bash_dequote_filename ((char *)text, qc);
+  else
+    dfn = (char *)text;
   m1 = rl_completion_matches (dfn, rl_filename_completion_function);
-  free (dfn);
+  if (dfn != text)
+    free (dfn);
 
   if (m1 == 0 || m1[0] == 0)
     return m1;
--- a/builtins/common.h	2013-07-08 22:54:47.000000000 +0200
+++ b/builtins/common.h	2014-10-11 10:34:38.174610956 +0200
@@ -33,6 +33,8 @@
 #define SEVAL_RESETLINE	0x010
 #define SEVAL_PARSEONLY	0x020
 #define SEVAL_NOLONGJMP 0x040
+#define SEVAL_FUNCDEF	0x080		/* only allow function definitions */
+#define SEVAL_ONECMD	0x100		/* only allow a single command */
 
 /* Flags for describe_command, shared between type.def and command.def */
 #define CDESC_ALL		0x001	/* type -a */
--- a/builtins/evalstring.c	2014-02-11 15:42:10.000000000 +0100
+++ b/builtins/evalstring.c	2014-10-11 10:37:09.625510026 +0200
@@ -308,6 +308,27 @@
 	    {
 	      struct fd_bitmap *bitmap;
 
+	      if (flags & SEVAL_FUNCDEF)
+		{
+		  char *x;
+
+		  /* If the command parses to something other than a straight
+		     function definition, or if we have not consumed the entire
+		     string, or if the parser has transformed the function
+		     name (as parsing will if it begins or ends with shell
+		     whitespace, for example), reject the attempt */
+		  if (command->type != cm_function_def ||
+		      ((x = parser_remaining_input ()) && *x) ||
+		      (STREQ (from_file, command->value.Function_def->name->word) == 0))
+		    {
+		      internal_warning (_("%s: ignoring function definition attempt"), from_file);
+		      should_jump_to_top_level = 0;
+		      last_result = last_command_exit_value = EX_BADUSAGE;
+		      reset_parser ();
+		      break;
+		    }
+		}
+
 	      bitmap = new_fd_bitmap (FD_BITMAP_SIZE);
 	      begin_unwind_frame ("pe_dispose");
 	      add_unwind_protect (dispose_fd_bitmap, bitmap);
@@ -368,6 +389,12 @@
 	      dispose_command (command);
 	      dispose_fd_bitmap (bitmap);
 	      discard_unwind_frame ("pe_dispose");
+
+	      if (flags & SEVAL_ONECMD)
+		{
+		  reset_parser ();
+		  break;
+		}
 	    }
 	}
       else
--- a/builtins/read.def	2013-09-02 17:54:00.000000000 +0200
+++ b/builtins/read.def	2014-10-11 10:34:38.174610956 +0200
@@ -442,7 +442,10 @@
       add_unwind_protect (reset_alarm, (char *)NULL);
 #if defined (READLINE)
       if (edit)
-	add_unwind_protect (reset_attempted_completion_function, (char *)NULL);
+	{
+	  add_unwind_protect (reset_attempted_completion_function, (char *)NULL);
+	  add_unwind_protect (bashline_reset_event_hook, (char *)NULL);
+	}
 #endif
       falarm (tmsec, tmusec);
     }
@@ -1021,6 +1024,7 @@
 
   old_attempted_completion_function = rl_attempted_completion_function;
   rl_attempted_completion_function = (rl_completion_func_t *)NULL;
+  bashline_set_event_hook ();
   if (itext)
     {
       old_startup_hook = rl_startup_hook;
@@ -1032,6 +1036,7 @@
 
   rl_attempted_completion_function = old_attempted_completion_function;
   old_attempted_completion_function = (rl_completion_func_t *)NULL;
+  bashline_reset_event_hook ();
 
   if (ret == 0)
     return ret;
--- a/copy_cmd.c	2009-09-11 22:28:02.000000000 +0200
+++ b/copy_cmd.c	2014-10-11 10:34:38.174610956 +0200
@@ -126,7 +126,7 @@
     {
     case r_reading_until:
     case r_deblank_reading_until:
-      new_redirect->here_doc_eof = savestring (redirect->here_doc_eof);
+      new_redirect->here_doc_eof = redirect->here_doc_eof ? savestring (redirect->here_doc_eof) : 0;
       /*FALLTHROUGH*/
     case r_reading_string:
     case r_appending_to:
--- a/execute_cmd.c	2014-01-31 16:54:52.000000000 +0100
+++ b/execute_cmd.c	2014-10-11 10:34:38.175610976 +0200
@@ -2409,7 +2409,16 @@
 #endif
       lstdin = wait_for (lastpid);
 #if defined (JOB_CONTROL)
-      exec_result = job_exit_status (lastpipe_jid);
+      /* If wait_for removes the job from the jobs table, use result of last
+	 command as pipeline's exit status as usual.  The jobs list can get
+	 frozen and unfrozen at inconvenient times if there are multiple pipelines
+	 running simultaneously. */
+      if (INVALID_JOB (lastpipe_jid) == 0)
+	exec_result = job_exit_status (lastpipe_jid);
+      else if (pipefail_opt)
+	exec_result = exec_result | lstdin;	/* XXX */
+      /* otherwise we use exec_result */
+        
 #endif
       unfreeze_jobs_list ();
     }
--- a/externs.h	2014-01-02 20:58:20.000000000 +0100
+++ b/externs.h	2014-10-11 10:34:38.175610976 +0200
@@ -324,6 +324,7 @@
 extern char *sh_backslash_quote __P((char *, const char *, int));
 extern char *sh_backslash_quote_for_double_quotes __P((char *));
 extern int sh_contains_shell_metas __P((char *));
+extern int sh_contains_quotes __P((char *));
 
 /* declarations for functions defined in lib/sh/spell.c */
 extern int spname __P((char *, char *));
--- a/jobs.c	2014-01-10 15:05:34.000000000 +0100
+++ b/jobs.c	2014-10-11 10:34:38.176610995 +0200
@@ -3597,6 +3597,7 @@
   unwind_protect_int (jobs_list_frozen);
   unwind_protect_pointer (the_pipeline);
   unwind_protect_pointer (subst_assign_varlist);
+  unwind_protect_pointer (this_shell_builtin);
 
   /* We have to add the commands this way because they will be run
      in reverse order of adding.  We don't want maybe_set_sigchld_trap ()
@@ -4374,7 +4375,7 @@
 void
 end_job_control ()
 {
-  if (interactive_shell)		/* XXX - should it be interactive? */
+  if (interactive_shell || job_control)		/* XXX - should it be just job_control? */
     {
       terminate_stopped_jobs ();
 
--- a/lib/glob/glob.c	2014-02-01 03:43:51.000000000 +0100
+++ b/lib/glob/glob.c	2014-10-11 10:34:38.176610995 +0200
@@ -123,6 +123,8 @@
 extern char *glob_patscan __P((char *, char *, int));
 extern wchar_t *glob_patscan_wc __P((wchar_t *, wchar_t *, int));
 
+extern char *glob_dirscan __P((char *, int));
+
 /* Compile `glob_loop.c' for single-byte characters. */
 #define CHAR	unsigned char
 #define INT	int
@@ -179,42 +181,53 @@
      char *pat, *dname;
      int flags;
 {
-  char *pp, *pe, *t;
-  int n, r;
+  char *pp, *pe, *t, *se;
+  int n, r, negate;
 
+  negate = *pat == '!';
   pp = pat + 2;
-  pe = pp + strlen (pp) - 1;	/*(*/
-  if (*pe != ')')
+  se = pp + strlen (pp) - 1;		/* end of string */
+  pe = glob_patscan (pp, se, 0);	/* end of extglob pattern (( */
+  /* we should check for invalid extglob pattern here */
+  if (pe == 0)
     return 0;
-  if ((t = strchr (pp, '|')) == 0)	/* easy case first */
+
+  /* if pe != se we have more of the pattern at the end of the extglob
+     pattern. Check the easy case first ( */
+  if (pe == se && *pe == ')' && (t = strchr (pp, '|')) == 0)
     {
       *pe = '\0';
+#if defined (HANDLE_MULTIBYTE)
+      r = mbskipname (pp, dname, flags);
+#else
       r = skipname (pp, dname, flags);	/*(*/
+#endif
       *pe = ')';
       return r;
     }
+
+  /* check every subpattern */
   while (t = glob_patscan (pp, pe, '|'))
     {
       n = t[-1];
       t[-1] = '\0';
+#if defined (HANDLE_MULTIBYTE)
+      r = mbskipname (pp, dname, flags);
+#else
       r = skipname (pp, dname, flags);
+#endif
       t[-1] = n;
       if (r == 0)	/* if any pattern says not skip, we don't skip */
         return r;
       pp = t;
     }	/*(*/
 
-  if (pp == pe)		/* glob_patscan might find end of pattern */
+  /* glob_patscan might find end of pattern */
+  if (pp == se)
     return r;
 
-  *pe = '\0';
-#  if defined (HANDLE_MULTIBYTE)
-  r = mbskipname (pp, dname, flags);	/*(*/
-#  else
-  r = skipname (pp, dname, flags);	/*(*/
-#  endif
-  *pe = ')';
-  return r;
+  /* but if it doesn't then we didn't match a leading dot */
+  return 0;
 }
 #endif
 
@@ -277,20 +290,23 @@
      int flags;
 {
 #if EXTENDED_GLOB
-  wchar_t *pp, *pe, *t, n;
-  int r;
+  wchar_t *pp, *pe, *t, n, *se;
+  int r, negate;
 
+  negate = *pat == L'!';
   pp = pat + 2;
-  pe = pp + wcslen (pp) - 1;	/*(*/
-  if (*pe != L')')
-    return 0;
-  if ((t = wcschr (pp, L'|')) == 0)
+  se = pp + wcslen (pp) - 1;	/*(*/
+  pe = glob_patscan_wc (pp, se, 0);
+
+  if (pe == se && *pe == ')' && (t = wcschr (pp, L'|')) == 0)
     {
       *pe = L'\0';
       r = wchkname (pp, dname); /*(*/
       *pe = L')';
       return r;
     }
+
+  /* check every subpattern */
   while (t = glob_patscan_wc (pp, pe, '|'))
     {
       n = t[-1];
@@ -305,10 +321,8 @@
   if (pp == pe)		/* glob_patscan_wc might find end of pattern */
     return r;
 
-  *pe = L'\0';
-  r = wchkname (pp, dname);	/*(*/
-  *pe = L')';
-  return r;
+  /* but if it doesn't then we didn't match a leading dot */
+  return 0;
 #else
   return (wchkname (pat, dname));
 #endif
@@ -1006,7 +1020,7 @@
 {
   char **result;
   unsigned int result_size;
-  char *directory_name, *filename, *dname;
+  char *directory_name, *filename, *dname, *fn;
   unsigned int directory_len;
   int free_dirname;			/* flag */
   int dflags;
@@ -1022,6 +1036,18 @@
 
   /* Find the filename.  */
   filename = strrchr (pathname, '/');
+#if defined (EXTENDED_GLOB)
+  if (filename && extended_glob)
+    {
+      fn = glob_dirscan (pathname, '/');
+#if DEBUG_MATCHING
+      if (fn != filename)
+	fprintf (stderr, "glob_filename: glob_dirscan: fn (%s) != filename (%s)\n", fn ? fn : "(null)", filename);
+#endif
+      filename = fn;
+    }
+#endif
+
   if (filename == NULL)
     {
       filename = pathname;
--- a/lib/glob/gmisc.c	2013-10-28 19:45:25.000000000 +0100
+++ b/lib/glob/gmisc.c	2014-10-11 10:34:38.176610995 +0200
@@ -42,6 +42,8 @@
 #define WLPAREN         L'('
 #define WRPAREN         L')'
 
+extern char *glob_patscan __P((char *, char *, int));
+
 /* Return 1 of the first character of WSTRING could match the first
    character of pattern WPAT.  Wide character version. */
 int
@@ -210,6 +212,7 @@
     case '+':
     case '!':
     case '@':
+    case '?':
       return (pat[1] == LPAREN);
     default:
       return 0;
@@ -374,3 +377,34 @@
 
   return matlen;
 }
+
+/* Skip characters in PAT and return the final occurrence of DIRSEP.  This
+   is only called when extended_glob is set, so we have to skip over extglob
+   patterns x(...) */
+char *
+glob_dirscan (pat, dirsep)
+     char *pat;
+     int dirsep;
+{
+  char *p, *d, *pe, *se;
+
+  d = pe = se = 0;
+  for (p = pat; p && *p; p++)
+    {
+      if (extglob_pattern_p (p))
+	{
+	  if (se == 0)
+	    se = p + strlen (p) - 1;
+	  pe = glob_patscan (p + 2, se, 0);
+	  if (pe == 0)
+	    continue;
+	  else if (*pe == 0)
+	    break;
+	  p = pe - 1;	/* will do increment above */
+	  continue;
+	}
+      if (*p ==  dirsep)
+	d = p;
+    }
+  return d;
+}
--- a/lib/readline/display.c	2013-12-27 19:10:56.000000000 +0100
+++ b/lib/readline/display.c	2014-10-11 10:34:38.177611015 +0200
@@ -1637,7 +1637,7 @@
   /* If we are changing the number of invisible characters in a line, and
      the spot of first difference is before the end of the invisible chars,
      lendiff needs to be adjusted. */
-  if (current_line == 0 && !_rl_horizontal_scroll_mode &&
+  if (current_line == 0 && /* !_rl_horizontal_scroll_mode && */
       current_invis_chars != visible_wrap_offset)
     {
       if (MB_CUR_MAX > 1 && rl_byte_oriented == 0)
@@ -1825,8 +1825,13 @@
 	      else
 		_rl_last_c_pos += bytes_to_insert;
 
+	      /* XXX - we only want to do this if we are at the end of the line
+		 so we move there with _rl_move_cursor_relative */
 	      if (_rl_horizontal_scroll_mode && ((oe-old) > (ne-new)))
-		goto clear_rest_of_line;
+		{
+		  _rl_move_cursor_relative (ne-new, new);
+		  goto clear_rest_of_line;
+		}
 	    }
 	}
       /* Otherwise, print over the existing material. */
@@ -2677,7 +2682,8 @@
 {
   if (_rl_echoing_p)
     {
-      _rl_move_vert (_rl_vis_botlin);
+      if (_rl_vis_botlin > 0)	/* minor optimization plus bug fix */
+	_rl_move_vert (_rl_vis_botlin);
       _rl_vis_botlin = 0;
       fflush (rl_outstream);
       rl_restart_output (1, 0);
--- a/lib/readline/input.c	2014-01-10 21:07:08.000000000 +0100
+++ b/lib/readline/input.c	2014-10-11 10:34:38.177611015 +0200
@@ -534,8 +534,16 @@
 	return (RL_ISSTATE (RL_STATE_READCMD) ? READERR : EOF);
       else if (_rl_caught_signal == SIGHUP || _rl_caught_signal == SIGTERM)
 	return (RL_ISSTATE (RL_STATE_READCMD) ? READERR : EOF);
+      /* keyboard-generated signals of interest */
       else if (_rl_caught_signal == SIGINT || _rl_caught_signal == SIGQUIT)
         RL_CHECK_SIGNALS ();
+      /* non-keyboard-generated signals of interest */
+      else if (_rl_caught_signal == SIGALRM
+#if defined (SIGVTALRM)
+		|| _rl_caught_signal == SIGVTALRM
+#endif
+	      )
+        RL_CHECK_SIGNALS ();
 
       if (rl_signal_event_hook)
 	(*rl_signal_event_hook) ();
--- a/lib/readline/misc.c	2012-09-02 00:03:11.000000000 +0200
+++ b/lib/readline/misc.c	2014-10-11 10:34:38.177611015 +0200
@@ -461,6 +461,7 @@
 	    saved_undo_list = 0;
 	  /* Set up rl_line_buffer and other variables from history entry */
 	  rl_replace_from_history (entry, 0);	/* entry->line is now current */
+	  entry->data = 0;			/* entry->data is now current undo list */
 	  /* Undo all changes to this history entry */
 	  while (rl_undo_list)
 	    rl_do_undo ();
@@ -468,7 +469,6 @@
 	     the timestamp. */
 	  FREE (entry->line);
 	  entry->line = savestring (rl_line_buffer);
-	  entry->data = 0;
 	}
       entry = previous_history ();
     }
--- a/lib/readline/readline.c	2013-10-28 19:58:06.000000000 +0100
+++ b/lib/readline/readline.c	2014-10-11 10:34:38.177611015 +0200
@@ -744,7 +744,8 @@
     r = _rl_subseq_result (r, cxt->oldmap, cxt->okey, (cxt->flags & KSEQ_SUBSEQ));
 
   RL_CHECK_SIGNALS ();
-  if (r == 0)			/* success! */
+  /* We only treat values < 0 specially to simulate recursion. */
+  if (r >= 0 || (r == -1 && (cxt->flags & KSEQ_SUBSEQ) == 0))	/* success! or failure! */
     {
       _rl_keyseq_chain_dispose ();
       RL_UNSETSTATE (RL_STATE_MULTIKEY);
@@ -964,7 +965,7 @@
 #if defined (VI_MODE)
   if (rl_editing_mode == vi_mode && _rl_keymap == vi_movement_keymap &&
       key != ANYOTHERKEY &&
-      rl_key_sequence_length == 1 &&	/* XXX */
+      _rl_dispatching_keymap == vi_movement_keymap &&
       _rl_vi_textmod_command (key))
     _rl_vi_set_last (key, rl_numeric_arg, rl_arg_sign);
 #endif
--- a/lib/sh/shquote.c	2013-04-01 03:53:32.000000000 +0200
+++ b/lib/sh/shquote.c	2014-10-11 10:34:38.177611015 +0200
@@ -311,3 +311,17 @@
 
   return (0);
 }
+
+int
+sh_contains_quotes (string)
+     char *string;
+{
+  char *s;
+
+  for (s = string; s && *s; s++)
+    {
+      if (*s == '\'' || *s == '"' || *s == '\\')
+	return 1;
+    }
+  return 0;
+}
--- a/make_cmd.c	2011-12-16 14:08:01.000000000 +0100
+++ b/make_cmd.c	2014-10-11 10:34:38.178611034 +0200
@@ -692,6 +692,7 @@
   /* First do the common cases. */
   temp->redirector = source;
   temp->redirectee = dest_and_filename;
+  temp->here_doc_eof = 0;
   temp->instruction = instruction;
   temp->flags = 0;
   temp->rflags = flags;
--- a/parse.y	2014-02-11 15:42:10.000000000 +0100
+++ b/parse.y	2014-10-11 10:37:09.626510045 +0200
@@ -168,6 +168,9 @@
 
 static int reserved_word_acceptable __P((int));
 static int yylex __P((void));
+
+static void push_heredoc __P((REDIRECT *));
+static char *mk_alexpansion __P((char *));
 static int alias_expand_token __P((char *));
 static int time_command_acceptable __P((void));
 static int special_case_tokens __P((char *));
@@ -265,7 +268,9 @@
 
 /* Variables to manage the task of reading here documents, because we need to
    defer the reading until after a complete command has been collected. */
-static REDIRECT *redir_stack[10];
+#define HEREDOC_MAX 16
+
+static REDIRECT *redir_stack[HEREDOC_MAX];
 int need_here_doc;
 
 /* Where shell input comes from.  History expansion is performed on each
@@ -307,7 +312,7 @@
    or `for WORD' begins.  This is a nested command maximum, since the array
    index is decremented after a case, select, or for command is parsed. */
 #define MAX_CASE_NEST	128
-static int word_lineno[MAX_CASE_NEST];
+static int word_lineno[MAX_CASE_NEST+1];
 static int word_top = -1;
 
 /* If non-zero, it is the token that we want read_token to return
@@ -520,42 +525,42 @@
 			  source.dest = 0;
 			  redir.filename = $2;
 			  $$ = make_redirection (source, r_reading_until, redir, 0);
-			  redir_stack[need_here_doc++] = $$;
+			  push_heredoc ($$);
 			}
 	|	NUMBER LESS_LESS WORD
 			{
 			  source.dest = $1;
 			  redir.filename = $3;
 			  $$ = make_redirection (source, r_reading_until, redir, 0);
-			  redir_stack[need_here_doc++] = $$;
+			  push_heredoc ($$);
 			}
 	|	REDIR_WORD LESS_LESS WORD
 			{
 			  source.filename = $1;
 			  redir.filename = $3;
 			  $$ = make_redirection (source, r_reading_until, redir, REDIR_VARASSIGN);
-			  redir_stack[need_here_doc++] = $$;
+			  push_heredoc ($$);
 			}
 	|	LESS_LESS_MINUS WORD
 			{
 			  source.dest = 0;
 			  redir.filename = $2;
 			  $$ = make_redirection (source, r_deblank_reading_until, redir, 0);
-			  redir_stack[need_here_doc++] = $$;
+			  push_heredoc ($$);
 			}
 	|	NUMBER LESS_LESS_MINUS WORD
 			{
 			  source.dest = $1;
 			  redir.filename = $3;
 			  $$ = make_redirection (source, r_deblank_reading_until, redir, 0);
-			  redir_stack[need_here_doc++] = $$;
+			  push_heredoc ($$);
 			}
 	|	REDIR_WORD  LESS_LESS_MINUS WORD
 			{
 			  source.filename = $1;
 			  redir.filename = $3;
 			  $$ = make_redirection (source, r_deblank_reading_until, redir, REDIR_VARASSIGN);
-			  redir_stack[need_here_doc++] = $$;
+			  push_heredoc ($$);
 			}
 	|	LESS_LESS_LESS WORD
 			{
@@ -2424,7 +2429,7 @@
 	 not already end in an EOF character.  */
       if (shell_input_line_terminator != EOF)
 	{
-	  if (shell_input_line_size < SIZE_MAX && shell_input_line_len > shell_input_line_size - 3)
+	  if (shell_input_line_size < SIZE_MAX-3 && (shell_input_line_len+3 > shell_input_line_size))
 	    shell_input_line = (char *)xrealloc (shell_input_line,
 					1 + (shell_input_line_size += 2));
 
@@ -2533,6 +2538,16 @@
     eol_ungetc_lookahead = c;
 }
 
+char *
+parser_remaining_input ()
+{
+  if (shell_input_line == 0)
+    return 0;
+  if (shell_input_line_index < 0 || shell_input_line_index >= shell_input_line_len)
+    return '\0';	/* XXX */
+  return (shell_input_line + shell_input_line_index);
+}
+
 #ifdef INCLUDE_UNUSED
 /* Back the input pointer up by one, effectively `ungetting' a character. */
 static void
@@ -2636,13 +2651,28 @@
    which allow ESAC to be the next one read. */
 static int esacs_needed_count;
 
+static void
+push_heredoc (r)
+     REDIRECT *r;
+{
+  if (need_here_doc >= HEREDOC_MAX)
+    {
+      last_command_exit_value = EX_BADUSAGE;
+      need_here_doc = 0;
+      report_syntax_error (_("maximum here-document count exceeded"));
+      reset_parser ();
+      exit_shell (last_command_exit_value);
+    }
+  redir_stack[need_here_doc++] = r;
+}
+
 void
 gather_here_documents ()
 {
   int r;
 
   r = 0;
-  while (need_here_doc)
+  while (need_here_doc > 0)
     {
       parser_state |= PST_HEREDOC;
       make_here_document (redir_stack[r++], line_number);
@@ -2953,6 +2983,8 @@
   FREE (word_desc_to_read);
   word_desc_to_read = (WORD_DESC *)NULL;
 
+  eol_ungetc_lookahead = 0;
+
   current_token = '\n';		/* XXX */
   last_read_token = '\n';
   token_to_read = '\n';
@@ -3398,7 +3430,7 @@
          within a double-quoted ${...} construct "an even number of
          unescaped double-quotes or single-quotes, if any, shall occur." */
       /* This was changed in Austin Group Interp 221 */
-      if MBTEST(posixly_correct && shell_compatibility_level > 41 && dolbrace_state != DOLBRACE_QUOTE && (flags & P_DQUOTE) && (flags & P_DOLBRACE) && ch == '\'')
+      if MBTEST(posixly_correct && shell_compatibility_level > 41 && dolbrace_state != DOLBRACE_QUOTE && dolbrace_state != DOLBRACE_QUOTE2 && (flags & P_DQUOTE) && (flags & P_DOLBRACE) && ch == '\'')
 	continue;
 
       /* Could also check open == '`' if we want to parse grouping constructs
@@ -4005,8 +4037,8 @@
   reset_parser ();
   /* reset_parser clears shell_input_line and associated variables */
   restore_input_line_state (&ls);
-  if (interactive)
-    token_to_read = 0;
+
+  token_to_read = 0;
 
   /* Need to find how many characters parse_and_execute consumed, update
      *indp, if flags != 0, copy the portion of the string parsed into RET
@@ -6075,6 +6107,7 @@
 
   ps->expand_aliases = expand_aliases;
   ps->echo_input_at_read = echo_input_at_read;
+  ps->need_here_doc = need_here_doc;
 
   ps->token = token;
   ps->token_buffer_size = token_buffer_size;
@@ -6123,6 +6156,7 @@
 
   expand_aliases = ps->expand_aliases;
   echo_input_at_read = ps->echo_input_at_read;
+  need_here_doc = ps->need_here_doc;
 
   FREE (token);
   token = ps->token;
--- a/patchlevel.h	2012-12-29 16:47:57.000000000 +0100
+++ b/patchlevel.h	2014-10-11 10:37:09.628510082 +0200
@@ -25,6 +25,6 @@
    regexp `^#define[ 	]*PATCHLEVEL', since that's what support/mkversion.sh
    looks for to find the patch level (for the sccs version string). */
 
-#define PATCHLEVEL 0
+#define PATCHLEVEL 30
 
 #endif /* _PATCHLEVEL_H_ */
--- a/pcomplete.c	2013-08-26 21:23:45.000000000 +0200
+++ b/pcomplete.c	2014-10-11 10:34:38.179611054 +0200
@@ -183,6 +183,7 @@
 
 COMPSPEC *pcomp_curcs;
 const char *pcomp_curcmd;
+const char *pcomp_curtxt;
 
 #ifdef DEBUG
 /* Debugging code */
@@ -753,6 +754,32 @@
 	     quoted strings. */
 	  dfn = (*rl_filename_dequoting_function) ((char *)text, rl_completion_quote_character);
 	}
+      /* Intended to solve a mismatched assumption by bash-completion.  If
+	 the text to be completed is empty, but bash-completion turns it into
+	 a quoted string ('') assuming that this code will dequote it before
+	 calling readline, do the dequoting. */
+      else if (iscompgen && iscompleting &&
+	       pcomp_curtxt && *pcomp_curtxt == 0 &&
+	       text && (*text == '\'' || *text == '"') && text[1] == text[0] && text[2] == 0 && 
+	       rl_filename_dequoting_function)
+	dfn = (*rl_filename_dequoting_function) ((char *)text, rl_completion_quote_character);
+      /* Another mismatched assumption by bash-completion.  If compgen is being
+      	 run as part of bash-completion, and the argument to compgen is not
+      	 the same as the word originally passed to the programmable completion
+      	 code, dequote the argument if it has quote characters.  It's an
+      	 attempt to detect when bash-completion is quoting its filename
+      	 argument before calling compgen. */
+      /* We could check whether gen_shell_function_matches is in the call
+	 stack by checking whether the gen-shell-function-matches tag is in
+	 the unwind-protect stack, but there's no function to do that yet.
+	 We could simply check whether we're executing in a function by
+	 checking variable_context, and may end up doing that. */
+      else if (iscompgen && iscompleting && rl_filename_dequoting_function &&
+	       pcomp_curtxt && text &&
+	       STREQ (pcomp_curtxt, text) == 0 &&
+	       variable_context &&
+	       sh_contains_quotes (text))	/* guess */
+	dfn = (*rl_filename_dequoting_function) ((char *)text, rl_completion_quote_character);
       else
 	dfn = savestring (text);
     }
@@ -1522,7 +1549,7 @@
      COMPSPEC **lastcs;
 {
   COMPSPEC *cs, *oldcs;
-  const char *oldcmd;
+  const char *oldcmd, *oldtxt;
   STRINGLIST *ret;
 
   cs = progcomp_search (ocmd);
@@ -1545,14 +1572,17 @@
 
   oldcs = pcomp_curcs;
   oldcmd = pcomp_curcmd;
+  oldtxt = pcomp_curtxt;
 
   pcomp_curcs = cs;
   pcomp_curcmd = cmd;
+  pcomp_curtxt = word;
 
   ret = gen_compspec_completions (cs, cmd, word, start, end, foundp);
 
   pcomp_curcs = oldcs;
   pcomp_curcmd = oldcmd;
+  pcomp_curtxt = oldtxt;
 
   /* We need to conditionally handle setting *retryp here */
   if (retryp)
--- a/shell.h	2012-12-26 03:11:01.000000000 +0100
+++ b/shell.h	2014-10-11 10:37:09.626510045 +0200
@@ -168,7 +168,8 @@
   /* flags state affecting the parser */
   int expand_aliases;
   int echo_input_at_read;
-  
+  int need_here_doc;
+
 } sh_parser_state_t;
 
 typedef struct _sh_input_line_state_t {
@@ -179,6 +180,8 @@
 } sh_input_line_state_t;
 
 /* Let's try declaring these here. */
+extern char *parser_remaining_input __P((void));
+
 extern sh_parser_state_t *save_parser_state __P((sh_parser_state_t *));
 extern void restore_parser_state __P((sh_parser_state_t *));
 
--- a/subst.c	2014-01-23 22:26:37.000000000 +0100
+++ b/subst.c	2014-10-11 10:34:38.180611073 +0200
@@ -1192,12 +1192,18 @@
    Start extracting at (SINDEX) as if we had just seen "<(".
    Make (SINDEX) get the position of the matching ")". */ /*))*/
 char *
-extract_process_subst (string, starter, sindex)
+extract_process_subst (string, starter, sindex, xflags)
      char *string;
      char *starter;
      int *sindex;
+     int xflags;
 {
+#if 0
   return (extract_delimited_string (string, sindex, starter, "(", ")", SX_COMMAND));
+#else
+  xflags |= (no_longjmp_on_fatal_error ? SX_NOLONGJMP : 0);
+  return (xparse_dolparen (string, string+*sindex, sindex, xflags));
+#endif
 }
 #endif /* PROCESS_SUBSTITUTION */
 
@@ -1785,7 +1791,7 @@
 	  si = i + 2;
 	  if (string[si] == '\0')
 	    CQ_RETURN(si);
-	  temp = extract_process_subst (string, (c == '<') ? "<(" : ">(", &si);
+	  temp = extract_process_subst (string, (c == '<') ? "<(" : ">(", &si, 0);
 	  free (temp);		/* no SX_ALLOC here */
 	  i = si;
 	  if (string[i] == '\0')
@@ -3248,8 +3254,10 @@
   if (w->word == 0 || w->word[0] == '\0')
     return ((char *)NULL);
 
+  expand_no_split_dollar_star = 1;
   w->flags |= W_NOSPLIT2;
   l = call_expand_word_internal (w, 0, 0, (int *)0, (int *)0);
+  expand_no_split_dollar_star = 0;
   if (l)
     {
       if (special == 0)			/* LHS */
@@ -7366,7 +7374,13 @@
     }
 
   if (want_indir)
-    tdesc = parameter_brace_expand_indir (name + 1, var_is_special, quoted, quoted_dollar_atp, contains_dollar_at);
+    {
+      tdesc = parameter_brace_expand_indir (name + 1, var_is_special, quoted, quoted_dollar_atp, contains_dollar_at);
+      /* Turn off the W_ARRAYIND flag because there is no way for this function
+	 to return the index we're supposed to be using. */
+      if (tdesc && tdesc->flags)
+	tdesc->flags &= ~W_ARRAYIND;
+    }
   else
     tdesc = parameter_brace_expand_word (name, var_is_special, quoted, PF_IGNUNBOUND|(pflags&(PF_NOSPLIT2|PF_ASSIGNRHS)), &ind);
 
@@ -7847,6 +7861,10 @@
 	 We also want to make sure that splitting is done no matter what --
 	 according to POSIX.2, this expands to a list of the positional
 	 parameters no matter what IFS is set to. */
+      /* XXX - what to do when in a context where word splitting is not
+	 performed? Even when IFS is not the default, posix seems to imply
+	 that we behave like unquoted $* ?  Maybe we should use PF_NOSPLIT2
+	 here. */
       temp = string_list_dollar_at (list, (pflags & PF_ASSIGNRHS) ? (quoted|Q_DOUBLE_QUOTES) : quoted);
 
       tflag |= W_DOLLARAT;
@@ -8029,7 +8047,9 @@
 
 	  goto return0;
 	}
-      else if (var = find_variable_last_nameref (temp1))
+      else if (var && (invisible_p (var) || var_isset (var) == 0))
+	temp = (char *)NULL;
+      else if ((var = find_variable_last_nameref (temp1)) && var_isset (var) && invisible_p (var) == 0)
 	{
 	  temp = nameref_cell (var);
 #if defined (ARRAY_VARS)
@@ -8243,7 +8263,7 @@
 	    else
 	      t_index = sindex + 1; /* skip past both '<' and LPAREN */
 
-	    temp1 = extract_process_subst (string, (c == '<') ? "<(" : ">(", &t_index); /*))*/
+	    temp1 = extract_process_subst (string, (c == '<') ? "<(" : ">(", &t_index, 0); /*))*/
 	    sindex = t_index;
 
 	    /* If the process substitution specification is `<()', we want to
@@ -8816,6 +8836,7 @@
   else
     {
       char *ifs_chars;
+      char *tstring;
 
       ifs_chars = (quoted_dollar_at || has_dollar_at) ? ifs_value : (char *)NULL;
 
@@ -8830,11 +8851,36 @@
 	 regardless of what else has happened to IFS since the expansion. */
       if (split_on_spaces)
 	list = list_string (istring, " ", 1);	/* XXX quoted == 1? */
+      /* If we have $@ (has_dollar_at != 0) and we are in a context where we
+	 don't want to split the result (W_NOSPLIT2), and we are not quoted,
+	 we have already separated the arguments with the first character of
+	 $IFS.  In this case, we want to return a list with a single word
+	 with the separator possibly replaced with a space (it's what other
+	 shells seem to do).
+	 quoted_dollar_at is internal to this function and is set if we are
+	 passed an argument that is unquoted (quoted == 0) but we encounter a
+	 double-quoted $@ while expanding it. */
+      else if (has_dollar_at && quoted_dollar_at == 0 && ifs_chars && quoted == 0 && (word->flags & W_NOSPLIT2))
+	{
+	  /* Only split and rejoin if we have to */
+	  if (*ifs_chars && *ifs_chars != ' ')
+	    {
+	      list = list_string (istring, *ifs_chars ? ifs_chars : " ", 1);
+	      tstring = string_list (list);
+	    }
+	  else
+	    tstring = istring;
+	  tword = make_bare_word (tstring);
+	  if (tstring != istring)
+	    free (tstring);
+	  goto set_word_flags;
+	}
       else if (has_dollar_at && ifs_chars)
 	list = list_string (istring, *ifs_chars ? ifs_chars : " ", 1);
       else
 	{
 	  tword = make_bare_word (istring);
+set_word_flags:
 	  if ((quoted & (Q_DOUBLE_QUOTES|Q_HERE_DOCUMENT)) || (quoted_state == WHOLLY_QUOTED))
 	    tword->flags |= W_QUOTED;
 	  if (word->flags & W_ASSIGNMENT)
--- a/subst.h	2014-01-12 03:02:27.000000000 +0100
+++ b/subst.h	2014-10-11 10:34:38.181611093 +0200
@@ -82,7 +82,7 @@
 /* Extract the <( or >( construct in STRING, and return a new string.
    Start extracting at (SINDEX) as if we had just seen "<(".
    Make (SINDEX) get the position just after the matching ")". */
-extern char *extract_process_subst __P((char *, char *, int *));
+extern char *extract_process_subst __P((char *, char *, int *, int));
 #endif /* PROCESS_SUBSTITUTION */
 
 /* Extract the name of the variable to bind to from the assignment string. */
--- a/test.c	2014-02-04 22:52:58.000000000 +0100
+++ b/test.c	2014-10-11 10:34:38.181611093 +0200
@@ -646,8 +646,8 @@
       return (v && invisible_p (v) == 0 && var_isset (v) ? TRUE : FALSE);
 
     case 'R':
-      v = find_variable (arg);
-      return (v && invisible_p (v) == 0 && var_isset (v) && nameref_p (v) ? TRUE : FALSE);
+      v = find_variable_noref (arg);
+      return ((v && invisible_p (v) == 0 && var_isset (v) && nameref_p (v)) ? TRUE : FALSE);
     }
 
   /* We can't actually get here, but this shuts up gcc. */
@@ -723,6 +723,7 @@
     case 'o': case 'p': case 'r': case 's': case 't':
     case 'u': case 'v': case 'w': case 'x': case 'z':
     case 'G': case 'L': case 'O': case 'S': case 'N':
+    case 'R':
       return (1);
     }
 
--- a/trap.c	2014-02-05 16:03:21.000000000 +0100
+++ b/trap.c	2014-10-11 10:34:38.181611093 +0200
@@ -920,7 +920,8 @@
       subst_assign_varlist = 0;
 
 #if defined (JOB_CONTROL)
-      save_pipeline (1);	/* XXX only provides one save level */
+      if (sig != DEBUG_TRAP)	/* run_debug_trap does this */
+	save_pipeline (1);	/* XXX only provides one save level */
 #endif
 
       /* If we're in a function, make sure return longjmps come here, too. */
@@ -940,7 +941,8 @@
       trap_exit_value = last_command_exit_value;
 
 #if defined (JOB_CONTROL)
-      restore_pipeline (1);
+      if (sig != DEBUG_TRAP)	/* run_debug_trap does this */
+	restore_pipeline (1);
 #endif
 
       subst_assign_varlist = save_subst_varlist;
--- a/variables.c	2014-02-14 17:55:12.000000000 +0100
+++ b/variables.c	2014-10-11 10:34:38.182611113 +0200
@@ -83,6 +83,11 @@
 
 #define ifsname(s)	((s)[0] == 'I' && (s)[1] == 'F' && (s)[2] == 'S' && (s)[3] == '\0')
 
+#define BASHFUNC_PREFIX		"BASH_FUNC_"
+#define BASHFUNC_PREFLEN	10	/* == strlen(BASHFUNC_PREFIX */
+#define BASHFUNC_SUFFIX		"%%"
+#define BASHFUNC_SUFFLEN	2	/* == strlen(BASHFUNC_SUFFIX) */
+
 extern char **environ;
 
 /* Variables used here and defined in other files. */
@@ -279,7 +284,7 @@
 static void propagate_temp_var __P((PTR_T));
 static void dispose_temporary_env __P((sh_free_func_t *));     
 
-static inline char *mk_env_string __P((const char *, const char *));
+static inline char *mk_env_string __P((const char *, const char *, int));
 static char **make_env_array_from_var_list __P((SHELL_VAR **));
 static char **make_var_export_array __P((VAR_CONTEXT *));
 static char **make_func_export_array __P((void));
@@ -349,24 +354,33 @@
 
       /* If exported function, define it now.  Don't import functions from
 	 the environment in privileged mode. */
-      if (privmode == 0 && read_but_dont_execute == 0 && STREQN ("() {", string, 4))
+      if (privmode == 0 && read_but_dont_execute == 0 && 
+          STREQN (BASHFUNC_PREFIX, name, BASHFUNC_PREFLEN) &&
+          STREQ (BASHFUNC_SUFFIX, name + char_index - BASHFUNC_SUFFLEN) &&
+	  STREQN ("() {", string, 4))
 	{
-	  string_length = strlen (string);
-	  temp_string = (char *)xmalloc (3 + string_length + char_index);
+	  size_t namelen;
+	  char *tname;		/* desired imported function name */
 
-	  strcpy (temp_string, name);
-	  temp_string[char_index] = ' ';
-	  strcpy (temp_string + char_index + 1, string);
+	  namelen = char_index - BASHFUNC_PREFLEN - BASHFUNC_SUFFLEN;
 
-	  if (posixly_correct == 0 || legal_identifier (name))
-	    parse_and_execute (temp_string, name, SEVAL_NONINT|SEVAL_NOHIST);
+	  tname = name + BASHFUNC_PREFLEN;	/* start of func name */
+	  tname[namelen] = '\0';		/* now tname == func name */
+
+	  string_length = strlen (string);
+	  temp_string = (char *)xmalloc (namelen + string_length + 2);
 
-	  /* Ancient backwards compatibility.  Old versions of bash exported
-	     functions like name()=() {...} */
-	  if (name[char_index - 1] == ')' && name[char_index - 2] == '(')
-	    name[char_index - 2] = '\0';
+	  memcpy (temp_string, tname, namelen);
+	  temp_string[namelen] = ' ';
+	  memcpy (temp_string + namelen + 1, string, string_length + 1);
+
+	  /* Don't import function names that are invalid identifiers from the
+	     environment, though we still allow them to be defined as shell
+	     variables. */
+	  if (absolute_program (tname) == 0 && (posixly_correct == 0 || legal_identifier (tname)))
+	    parse_and_execute (temp_string, tname, SEVAL_NONINT|SEVAL_NOHIST|SEVAL_FUNCDEF|SEVAL_ONECMD);
 
-	  if (temp_var = find_function (name))
+	  if (temp_var = find_function (tname))
 	    {
 	      VSETATTR (temp_var, (att_exported|att_imported));
 	      array_needs_making = 1;
@@ -379,12 +393,11 @@
 		  array_needs_making = 1;
 		}
 	      last_command_exit_value = 1;
-	      report_error (_("error importing function definition for `%s'"), name);
+	      report_error (_("error importing function definition for `%s'"), tname);
 	    }
 
-	  /* ( */
-	  if (name[char_index - 1] == ')' && name[char_index - 2] == '\0')
-	    name[char_index - 2] = '(';		/* ) */
+	  /* Restore original suffix */
+	  tname[namelen] = BASHFUNC_SUFFIX[0];
 	}
 #if defined (ARRAY_VARS)
 #  if ARRAY_EXPORT
@@ -2197,10 +2210,7 @@
   /* local foo; local foo;  is a no-op. */
   old_var = find_variable (name);
   if (old_var && local_p (old_var) && old_var->context == variable_context)
-    {
-      VUNSETATTR (old_var, att_invisible);	/* XXX */
-      return (old_var);
-    }
+    return (old_var);
 
   was_tmpvar = old_var && tempvar_p (old_var);
   /* If we're making a local variable in a shell function, the temporary env
@@ -2963,7 +2973,7 @@
   var->context = variable_context;	/* XXX */
 
   INVALIDATE_EXPORTSTR (var);
-  var->exportstr = mk_env_string (name, value);
+  var->exportstr = mk_env_string (name, value, 0);
 
   array_needs_making = 1;
 
@@ -3861,21 +3871,42 @@
 /* **************************************************************** */
 
 static inline char *
-mk_env_string (name, value)
+mk_env_string (name, value, isfunc)
      const char *name, *value;
+     int isfunc;
 {
-  int name_len, value_len;
-  char	*p;
+  size_t name_len, value_len;
+  char	*p, *q;
 
   name_len = strlen (name);
   value_len = STRLEN (value);
-  p = (char *)xmalloc (2 + name_len + value_len);
-  strcpy (p, name);
-  p[name_len] = '=';
+
+  /* If we are exporting a shell function, construct the encoded function
+     name. */
+  if (isfunc && value)
+    {
+      p = (char *)xmalloc (BASHFUNC_PREFLEN + name_len + BASHFUNC_SUFFLEN + value_len + 2);
+      q = p;
+      memcpy (q, BASHFUNC_PREFIX, BASHFUNC_PREFLEN);
+      q += BASHFUNC_PREFLEN;
+      memcpy (q, name, name_len);
+      q += name_len;
+      memcpy (q, BASHFUNC_SUFFIX, BASHFUNC_SUFFLEN);
+      q += BASHFUNC_SUFFLEN;
+    }
+  else
+    {
+      p = (char *)xmalloc (2 + name_len + value_len);
+      memcpy (p, name, name_len);
+      q = p + name_len;
+    }
+
+  q[0] = '=';
   if (value && *value)
-    strcpy (p + name_len + 1, value);
+    memcpy (q + 1, value, value_len + 1);
   else
-    p[name_len + 1] = '\0';
+    q[1] = '\0';
+
   return (p);
 }
 
@@ -3961,7 +3992,7 @@
 	  /* Gee, I'd like to get away with not using savestring() if we're
 	     using the cached exportstr... */
 	  list[list_index] = USE_EXPORTSTR ? savestring (value)
-					   : mk_env_string (var->name, value);
+					   : mk_env_string (var->name, value, function_p (var));
 
 	  if (USE_EXPORTSTR == 0)
 	    SAVE_EXPORTSTR (var, list[list_index]);
--- a/y.tab.c	2014-02-11 16:57:47.000000000 +0100
+++ b/y.tab.c	2014-10-11 10:37:09.628510082 +0200
@@ -168,7 +168,7 @@
 
 
 /* Copy the first part of user declarations.  */
-#line 21 "/usr/homes/chet/src/bash/src/parse.y"
+#line 21 "/usr/src/local/bash/bash-4.3-patched/parse.y"
 
 #include "config.h"
 
@@ -319,6 +319,9 @@
 
 static int reserved_word_acceptable __P((int));
 static int yylex __P((void));
+
+static void push_heredoc __P((REDIRECT *));
+static char *mk_alexpansion __P((char *));
 static int alias_expand_token __P((char *));
 static int time_command_acceptable __P((void));
 static int special_case_tokens __P((char *));
@@ -416,7 +419,9 @@
 
 /* Variables to manage the task of reading here documents, because we need to
    defer the reading until after a complete command has been collected. */
-static REDIRECT *redir_stack[10];
+#define HEREDOC_MAX 16
+
+static REDIRECT *redir_stack[HEREDOC_MAX];
 int need_here_doc;
 
 /* Where shell input comes from.  History expansion is performed on each
@@ -458,7 +463,7 @@
    or `for WORD' begins.  This is a nested command maximum, since the array
    index is decremented after a case, select, or for command is parsed. */
 #define MAX_CASE_NEST	128
-static int word_lineno[MAX_CASE_NEST];
+static int word_lineno[MAX_CASE_NEST+1];
 static int word_top = -1;
 
 /* If non-zero, it is the token that we want read_token to return
@@ -492,7 +497,7 @@
 
 #if ! defined YYSTYPE && ! defined YYSTYPE_IS_DECLARED
 typedef union YYSTYPE
-#line 324 "/usr/homes/chet/src/bash/src/parse.y"
+#line 329 "/usr/src/local/bash/bash-4.3-patched/parse.y"
 {
   WORD_DESC *word;		/* the word that we read. */
   int number;			/* the number that we read. */
@@ -503,7 +508,7 @@
   PATTERN_LIST *pattern;
 }
 /* Line 193 of yacc.c.  */
-#line 507 "y.tab.c"
+#line 512 "y.tab.c"
 	YYSTYPE;
 # define yystype YYSTYPE /* obsolescent; will be withdrawn */
 # define YYSTYPE_IS_DECLARED 1
@@ -516,7 +521,7 @@
 
 
 /* Line 216 of yacc.c.  */
-#line 520 "y.tab.c"
+#line 525 "y.tab.c"
 
 #ifdef short
 # undef short
@@ -886,23 +891,23 @@
 /* YYRLINE[YYN] -- source line where rule number YYN was defined.  */
 static const yytype_uint16 yyrline[] =
 {
-       0,   377,   377,   388,   397,   412,   422,   424,   428,   434,
-     440,   446,   452,   458,   464,   470,   476,   482,   488,   494,
-     500,   506,   512,   518,   525,   532,   539,   546,   553,   560,
-     566,   572,   578,   584,   590,   596,   602,   608,   614,   620,
-     626,   632,   638,   644,   650,   656,   662,   668,   674,   680,
-     686,   692,   700,   702,   704,   708,   712,   723,   725,   729,
-     731,   733,   749,   751,   755,   757,   759,   761,   763,   765,
-     767,   769,   771,   773,   775,   779,   784,   789,   794,   799,
-     804,   809,   814,   821,   826,   831,   836,   843,   848,   853,
-     858,   863,   868,   875,   880,   885,   892,   895,   898,   902,
-     904,   935,   942,   947,   964,   969,   986,   993,   995,   997,
-    1002,  1006,  1010,  1014,  1016,  1018,  1022,  1023,  1027,  1029,
-    1031,  1033,  1037,  1039,  1041,  1043,  1045,  1047,  1051,  1053,
-    1062,  1070,  1071,  1077,  1078,  1085,  1089,  1091,  1093,  1100,
-    1102,  1104,  1108,  1109,  1112,  1114,  1116,  1120,  1121,  1130,
-    1143,  1159,  1174,  1176,  1178,  1185,  1188,  1192,  1194,  1200,
-    1206,  1223,  1243,  1245,  1268,  1272,  1274,  1276
+       0,   382,   382,   393,   402,   417,   427,   429,   433,   439,
+     445,   451,   457,   463,   469,   475,   481,   487,   493,   499,
+     505,   511,   517,   523,   530,   537,   544,   551,   558,   565,
+     571,   577,   583,   589,   595,   601,   607,   613,   619,   625,
+     631,   637,   643,   649,   655,   661,   667,   673,   679,   685,
+     691,   697,   705,   707,   709,   713,   717,   728,   730,   734,
+     736,   738,   754,   756,   760,   762,   764,   766,   768,   770,
+     772,   774,   776,   778,   780,   784,   789,   794,   799,   804,
+     809,   814,   819,   826,   831,   836,   841,   848,   853,   858,
+     863,   868,   873,   880,   885,   890,   897,   900,   903,   907,
+     909,   940,   947,   952,   969,   974,   991,   998,  1000,  1002,
+    1007,  1011,  1015,  1019,  1021,  1023,  1027,  1028,  1032,  1034,
+    1036,  1038,  1042,  1044,  1046,  1048,  1050,  1052,  1056,  1058,
+    1067,  1075,  1076,  1082,  1083,  1090,  1094,  1096,  1098,  1105,
+    1107,  1109,  1113,  1114,  1117,  1119,  1121,  1125,  1126,  1135,
+    1148,  1164,  1179,  1181,  1183,  1190,  1193,  1197,  1199,  1205,
+    1211,  1228,  1248,  1250,  1273,  1277,  1279,  1281
 };
 #endif
 
@@ -2093,7 +2098,7 @@
   switch (yyn)
     {
         case 2:
-#line 378 "/usr/homes/chet/src/bash/src/parse.y"
+#line 383 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  /* Case of regular command.  Discard the error
 			     safety net,and return the command just parsed. */
@@ -2107,7 +2112,7 @@
     break;
 
   case 3:
-#line 389 "/usr/homes/chet/src/bash/src/parse.y"
+#line 394 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  /* Case of regular command, but not a very
 			     interesting one.  Return a NULL command. */
@@ -2119,7 +2124,7 @@
     break;
 
   case 4:
-#line 398 "/usr/homes/chet/src/bash/src/parse.y"
+#line 403 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  /* Error during parsing.  Return NULL command. */
 			  global_command = (COMMAND *)NULL;
@@ -2137,7 +2142,7 @@
     break;
 
   case 5:
-#line 413 "/usr/homes/chet/src/bash/src/parse.y"
+#line 418 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  /* Case of EOF seen by itself.  Do ignoreeof or
 			     not. */
@@ -2148,17 +2153,17 @@
     break;
 
   case 6:
-#line 423 "/usr/homes/chet/src/bash/src/parse.y"
+#line 428 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.word_list) = make_word_list ((yyvsp[(1) - (1)].word), (WORD_LIST *)NULL); }
     break;
 
   case 7:
-#line 425 "/usr/homes/chet/src/bash/src/parse.y"
+#line 430 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.word_list) = make_word_list ((yyvsp[(2) - (2)].word), (yyvsp[(1) - (2)].word_list)); }
     break;
 
   case 8:
-#line 429 "/usr/homes/chet/src/bash/src/parse.y"
+#line 434 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 1;
 			  redir.filename = (yyvsp[(2) - (2)].word);
@@ -2167,7 +2172,7 @@
     break;
 
   case 9:
-#line 435 "/usr/homes/chet/src/bash/src/parse.y"
+#line 440 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 0;
 			  redir.filename = (yyvsp[(2) - (2)].word);
@@ -2176,7 +2181,7 @@
     break;
 
   case 10:
-#line 441 "/usr/homes/chet/src/bash/src/parse.y"
+#line 446 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2185,7 +2190,7 @@
     break;
 
   case 11:
-#line 447 "/usr/homes/chet/src/bash/src/parse.y"
+#line 452 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2194,7 +2199,7 @@
     break;
 
   case 12:
-#line 453 "/usr/homes/chet/src/bash/src/parse.y"
+#line 458 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2203,7 +2208,7 @@
     break;
 
   case 13:
-#line 459 "/usr/homes/chet/src/bash/src/parse.y"
+#line 464 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2212,7 +2217,7 @@
     break;
 
   case 14:
-#line 465 "/usr/homes/chet/src/bash/src/parse.y"
+#line 470 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 1;
 			  redir.filename = (yyvsp[(2) - (2)].word);
@@ -2221,7 +2226,7 @@
     break;
 
   case 15:
-#line 471 "/usr/homes/chet/src/bash/src/parse.y"
+#line 476 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2230,7 +2235,7 @@
     break;
 
   case 16:
-#line 477 "/usr/homes/chet/src/bash/src/parse.y"
+#line 482 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2239,7 +2244,7 @@
     break;
 
   case 17:
-#line 483 "/usr/homes/chet/src/bash/src/parse.y"
+#line 488 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 1;
 			  redir.filename = (yyvsp[(2) - (2)].word);
@@ -2248,7 +2253,7 @@
     break;
 
   case 18:
-#line 489 "/usr/homes/chet/src/bash/src/parse.y"
+#line 494 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2257,7 +2262,7 @@
     break;
 
   case 19:
-#line 495 "/usr/homes/chet/src/bash/src/parse.y"
+#line 500 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2266,7 +2271,7 @@
     break;
 
   case 20:
-#line 501 "/usr/homes/chet/src/bash/src/parse.y"
+#line 506 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 0;
 			  redir.filename = (yyvsp[(2) - (2)].word);
@@ -2275,7 +2280,7 @@
     break;
 
   case 21:
-#line 507 "/usr/homes/chet/src/bash/src/parse.y"
+#line 512 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2284,7 +2289,7 @@
     break;
 
   case 22:
-#line 513 "/usr/homes/chet/src/bash/src/parse.y"
+#line 518 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2293,67 +2298,67 @@
     break;
 
   case 23:
-#line 519 "/usr/homes/chet/src/bash/src/parse.y"
+#line 524 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 0;
 			  redir.filename = (yyvsp[(2) - (2)].word);
 			  (yyval.redirect) = make_redirection (source, r_reading_until, redir, 0);
-			  redir_stack[need_here_doc++] = (yyval.redirect);
+			  push_heredoc ((yyval.redirect));
 			}
     break;
 
   case 24:
-#line 526 "/usr/homes/chet/src/bash/src/parse.y"
+#line 531 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.filename = (yyvsp[(3) - (3)].word);
 			  (yyval.redirect) = make_redirection (source, r_reading_until, redir, 0);
-			  redir_stack[need_here_doc++] = (yyval.redirect);
+			  push_heredoc ((yyval.redirect));
 			}
     break;
 
   case 25:
-#line 533 "/usr/homes/chet/src/bash/src/parse.y"
+#line 538 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.filename = (yyvsp[(3) - (3)].word);
 			  (yyval.redirect) = make_redirection (source, r_reading_until, redir, REDIR_VARASSIGN);
-			  redir_stack[need_here_doc++] = (yyval.redirect);
+			  push_heredoc ((yyval.redirect));
 			}
     break;
 
   case 26:
-#line 540 "/usr/homes/chet/src/bash/src/parse.y"
+#line 545 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 0;
 			  redir.filename = (yyvsp[(2) - (2)].word);
 			  (yyval.redirect) = make_redirection (source, r_deblank_reading_until, redir, 0);
-			  redir_stack[need_here_doc++] = (yyval.redirect);
+			  push_heredoc ((yyval.redirect));
 			}
     break;
 
   case 27:
-#line 547 "/usr/homes/chet/src/bash/src/parse.y"
+#line 552 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.filename = (yyvsp[(3) - (3)].word);
 			  (yyval.redirect) = make_redirection (source, r_deblank_reading_until, redir, 0);
-			  redir_stack[need_here_doc++] = (yyval.redirect);
+			  push_heredoc ((yyval.redirect));
 			}
     break;
 
   case 28:
-#line 554 "/usr/homes/chet/src/bash/src/parse.y"
+#line 559 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.filename = (yyvsp[(3) - (3)].word);
 			  (yyval.redirect) = make_redirection (source, r_deblank_reading_until, redir, REDIR_VARASSIGN);
-			  redir_stack[need_here_doc++] = (yyval.redirect);
+			  push_heredoc ((yyval.redirect));
 			}
     break;
 
   case 29:
-#line 561 "/usr/homes/chet/src/bash/src/parse.y"
+#line 566 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 0;
 			  redir.filename = (yyvsp[(2) - (2)].word);
@@ -2362,7 +2367,7 @@
     break;
 
   case 30:
-#line 567 "/usr/homes/chet/src/bash/src/parse.y"
+#line 572 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2371,7 +2376,7 @@
     break;
 
   case 31:
-#line 573 "/usr/homes/chet/src/bash/src/parse.y"
+#line 578 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2380,7 +2385,7 @@
     break;
 
   case 32:
-#line 579 "/usr/homes/chet/src/bash/src/parse.y"
+#line 584 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 0;
 			  redir.dest = (yyvsp[(2) - (2)].number);
@@ -2389,7 +2394,7 @@
     break;
 
   case 33:
-#line 585 "/usr/homes/chet/src/bash/src/parse.y"
+#line 590 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.dest = (yyvsp[(3) - (3)].number);
@@ -2398,7 +2403,7 @@
     break;
 
   case 34:
-#line 591 "/usr/homes/chet/src/bash/src/parse.y"
+#line 596 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.dest = (yyvsp[(3) - (3)].number);
@@ -2407,7 +2412,7 @@
     break;
 
   case 35:
-#line 597 "/usr/homes/chet/src/bash/src/parse.y"
+#line 602 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 1;
 			  redir.dest = (yyvsp[(2) - (2)].number);
@@ -2416,7 +2421,7 @@
     break;
 
   case 36:
-#line 603 "/usr/homes/chet/src/bash/src/parse.y"
+#line 608 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.dest = (yyvsp[(3) - (3)].number);
@@ -2425,7 +2430,7 @@
     break;
 
   case 37:
-#line 609 "/usr/homes/chet/src/bash/src/parse.y"
+#line 614 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.dest = (yyvsp[(3) - (3)].number);
@@ -2434,7 +2439,7 @@
     break;
 
   case 38:
-#line 615 "/usr/homes/chet/src/bash/src/parse.y"
+#line 620 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 0;
 			  redir.filename = (yyvsp[(2) - (2)].word);
@@ -2443,7 +2448,7 @@
     break;
 
   case 39:
-#line 621 "/usr/homes/chet/src/bash/src/parse.y"
+#line 626 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2452,7 +2457,7 @@
     break;
 
   case 40:
-#line 627 "/usr/homes/chet/src/bash/src/parse.y"
+#line 632 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2461,7 +2466,7 @@
     break;
 
   case 41:
-#line 633 "/usr/homes/chet/src/bash/src/parse.y"
+#line 638 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 1;
 			  redir.filename = (yyvsp[(2) - (2)].word);
@@ -2470,7 +2475,7 @@
     break;
 
   case 42:
-#line 639 "/usr/homes/chet/src/bash/src/parse.y"
+#line 644 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2479,7 +2484,7 @@
     break;
 
   case 43:
-#line 645 "/usr/homes/chet/src/bash/src/parse.y"
+#line 650 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.filename = (yyvsp[(3) - (3)].word);
@@ -2488,7 +2493,7 @@
     break;
 
   case 44:
-#line 651 "/usr/homes/chet/src/bash/src/parse.y"
+#line 656 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 1;
 			  redir.dest = 0;
@@ -2497,7 +2502,7 @@
     break;
 
   case 45:
-#line 657 "/usr/homes/chet/src/bash/src/parse.y"
+#line 662 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.dest = 0;
@@ -2506,7 +2511,7 @@
     break;
 
   case 46:
-#line 663 "/usr/homes/chet/src/bash/src/parse.y"
+#line 668 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.dest = 0;
@@ -2515,7 +2520,7 @@
     break;
 
   case 47:
-#line 669 "/usr/homes/chet/src/bash/src/parse.y"
+#line 674 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 0;
 			  redir.dest = 0;
@@ -2524,7 +2529,7 @@
     break;
 
   case 48:
-#line 675 "/usr/homes/chet/src/bash/src/parse.y"
+#line 680 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = (yyvsp[(1) - (3)].number);
 			  redir.dest = 0;
@@ -2533,7 +2538,7 @@
     break;
 
   case 49:
-#line 681 "/usr/homes/chet/src/bash/src/parse.y"
+#line 686 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.filename = (yyvsp[(1) - (3)].word);
 			  redir.dest = 0;
@@ -2542,7 +2547,7 @@
     break;
 
   case 50:
-#line 687 "/usr/homes/chet/src/bash/src/parse.y"
+#line 692 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 1;
 			  redir.filename = (yyvsp[(2) - (2)].word);
@@ -2551,7 +2556,7 @@
     break;
 
   case 51:
-#line 693 "/usr/homes/chet/src/bash/src/parse.y"
+#line 698 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  source.dest = 1;
 			  redir.filename = (yyvsp[(2) - (2)].word);
@@ -2560,29 +2565,29 @@
     break;
 
   case 52:
-#line 701 "/usr/homes/chet/src/bash/src/parse.y"
+#line 706 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.element).word = (yyvsp[(1) - (1)].word); (yyval.element).redirect = 0; }
     break;
 
   case 53:
-#line 703 "/usr/homes/chet/src/bash/src/parse.y"
+#line 708 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.element).word = (yyvsp[(1) - (1)].word); (yyval.element).redirect = 0; }
     break;
 
   case 54:
-#line 705 "/usr/homes/chet/src/bash/src/parse.y"
+#line 710 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.element).redirect = (yyvsp[(1) - (1)].redirect); (yyval.element).word = 0; }
     break;
 
   case 55:
-#line 709 "/usr/homes/chet/src/bash/src/parse.y"
+#line 714 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.redirect) = (yyvsp[(1) - (1)].redirect);
 			}
     break;
 
   case 56:
-#line 713 "/usr/homes/chet/src/bash/src/parse.y"
+#line 718 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  register REDIRECT *t;
 
@@ -2594,27 +2599,27 @@
     break;
 
   case 57:
-#line 724 "/usr/homes/chet/src/bash/src/parse.y"
+#line 729 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_simple_command ((yyvsp[(1) - (1)].element), (COMMAND *)NULL); }
     break;
 
   case 58:
-#line 726 "/usr/homes/chet/src/bash/src/parse.y"
+#line 731 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_simple_command ((yyvsp[(2) - (2)].element), (yyvsp[(1) - (2)].command)); }
     break;
 
   case 59:
-#line 730 "/usr/homes/chet/src/bash/src/parse.y"
+#line 735 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = clean_simple_command ((yyvsp[(1) - (1)].command)); }
     break;
 
   case 60:
-#line 732 "/usr/homes/chet/src/bash/src/parse.y"
+#line 737 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 61:
-#line 734 "/usr/homes/chet/src/bash/src/parse.y"
+#line 739 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  COMMAND *tc;
 
@@ -2633,72 +2638,72 @@
     break;
 
   case 62:
-#line 750 "/usr/homes/chet/src/bash/src/parse.y"
+#line 755 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 63:
-#line 752 "/usr/homes/chet/src/bash/src/parse.y"
+#line 757 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 64:
-#line 756 "/usr/homes/chet/src/bash/src/parse.y"
+#line 761 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 65:
-#line 758 "/usr/homes/chet/src/bash/src/parse.y"
+#line 763 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 66:
-#line 760 "/usr/homes/chet/src/bash/src/parse.y"
+#line 765 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_while_command ((yyvsp[(2) - (5)].command), (yyvsp[(4) - (5)].command)); }
     break;
 
   case 67:
-#line 762 "/usr/homes/chet/src/bash/src/parse.y"
+#line 767 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_until_command ((yyvsp[(2) - (5)].command), (yyvsp[(4) - (5)].command)); }
     break;
 
   case 68:
-#line 764 "/usr/homes/chet/src/bash/src/parse.y"
+#line 769 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 69:
-#line 766 "/usr/homes/chet/src/bash/src/parse.y"
+#line 771 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 70:
-#line 768 "/usr/homes/chet/src/bash/src/parse.y"
+#line 773 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 71:
-#line 770 "/usr/homes/chet/src/bash/src/parse.y"
+#line 775 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 72:
-#line 772 "/usr/homes/chet/src/bash/src/parse.y"
+#line 777 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 73:
-#line 774 "/usr/homes/chet/src/bash/src/parse.y"
+#line 779 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 74:
-#line 776 "/usr/homes/chet/src/bash/src/parse.y"
+#line 781 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 75:
-#line 780 "/usr/homes/chet/src/bash/src/parse.y"
+#line 785 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_for_command ((yyvsp[(2) - (6)].word), add_string_to_list ("\"$@\"", (WORD_LIST *)NULL), (yyvsp[(5) - (6)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2706,7 +2711,7 @@
     break;
 
   case 76:
-#line 785 "/usr/homes/chet/src/bash/src/parse.y"
+#line 790 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_for_command ((yyvsp[(2) - (6)].word), add_string_to_list ("\"$@\"", (WORD_LIST *)NULL), (yyvsp[(5) - (6)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2714,7 +2719,7 @@
     break;
 
   case 77:
-#line 790 "/usr/homes/chet/src/bash/src/parse.y"
+#line 795 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_for_command ((yyvsp[(2) - (7)].word), add_string_to_list ("\"$@\"", (WORD_LIST *)NULL), (yyvsp[(6) - (7)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2722,7 +2727,7 @@
     break;
 
   case 78:
-#line 795 "/usr/homes/chet/src/bash/src/parse.y"
+#line 800 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_for_command ((yyvsp[(2) - (7)].word), add_string_to_list ("\"$@\"", (WORD_LIST *)NULL), (yyvsp[(6) - (7)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2730,7 +2735,7 @@
     break;
 
   case 79:
-#line 800 "/usr/homes/chet/src/bash/src/parse.y"
+#line 805 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_for_command ((yyvsp[(2) - (10)].word), REVERSE_LIST ((yyvsp[(5) - (10)].word_list), WORD_LIST *), (yyvsp[(9) - (10)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2738,7 +2743,7 @@
     break;
 
   case 80:
-#line 805 "/usr/homes/chet/src/bash/src/parse.y"
+#line 810 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_for_command ((yyvsp[(2) - (10)].word), REVERSE_LIST ((yyvsp[(5) - (10)].word_list), WORD_LIST *), (yyvsp[(9) - (10)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2746,7 +2751,7 @@
     break;
 
   case 81:
-#line 810 "/usr/homes/chet/src/bash/src/parse.y"
+#line 815 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_for_command ((yyvsp[(2) - (9)].word), (WORD_LIST *)NULL, (yyvsp[(8) - (9)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2754,7 +2759,7 @@
     break;
 
   case 82:
-#line 815 "/usr/homes/chet/src/bash/src/parse.y"
+#line 820 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_for_command ((yyvsp[(2) - (9)].word), (WORD_LIST *)NULL, (yyvsp[(8) - (9)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2762,7 +2767,7 @@
     break;
 
   case 83:
-#line 822 "/usr/homes/chet/src/bash/src/parse.y"
+#line 827 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 				  (yyval.command) = make_arith_for_command ((yyvsp[(2) - (7)].word_list), (yyvsp[(6) - (7)].command), arith_for_lineno);
 				  if (word_top > 0) word_top--;
@@ -2770,7 +2775,7 @@
     break;
 
   case 84:
-#line 827 "/usr/homes/chet/src/bash/src/parse.y"
+#line 832 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 				  (yyval.command) = make_arith_for_command ((yyvsp[(2) - (7)].word_list), (yyvsp[(6) - (7)].command), arith_for_lineno);
 				  if (word_top > 0) word_top--;
@@ -2778,7 +2783,7 @@
     break;
 
   case 85:
-#line 832 "/usr/homes/chet/src/bash/src/parse.y"
+#line 837 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 				  (yyval.command) = make_arith_for_command ((yyvsp[(2) - (5)].word_list), (yyvsp[(4) - (5)].command), arith_for_lineno);
 				  if (word_top > 0) word_top--;
@@ -2786,7 +2791,7 @@
     break;
 
   case 86:
-#line 837 "/usr/homes/chet/src/bash/src/parse.y"
+#line 842 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 				  (yyval.command) = make_arith_for_command ((yyvsp[(2) - (5)].word_list), (yyvsp[(4) - (5)].command), arith_for_lineno);
 				  if (word_top > 0) word_top--;
@@ -2794,7 +2799,7 @@
     break;
 
   case 87:
-#line 844 "/usr/homes/chet/src/bash/src/parse.y"
+#line 849 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_select_command ((yyvsp[(2) - (6)].word), add_string_to_list ("\"$@\"", (WORD_LIST *)NULL), (yyvsp[(5) - (6)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2802,7 +2807,7 @@
     break;
 
   case 88:
-#line 849 "/usr/homes/chet/src/bash/src/parse.y"
+#line 854 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_select_command ((yyvsp[(2) - (6)].word), add_string_to_list ("\"$@\"", (WORD_LIST *)NULL), (yyvsp[(5) - (6)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2810,7 +2815,7 @@
     break;
 
   case 89:
-#line 854 "/usr/homes/chet/src/bash/src/parse.y"
+#line 859 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_select_command ((yyvsp[(2) - (7)].word), add_string_to_list ("\"$@\"", (WORD_LIST *)NULL), (yyvsp[(6) - (7)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2818,7 +2823,7 @@
     break;
 
   case 90:
-#line 859 "/usr/homes/chet/src/bash/src/parse.y"
+#line 864 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_select_command ((yyvsp[(2) - (7)].word), add_string_to_list ("\"$@\"", (WORD_LIST *)NULL), (yyvsp[(6) - (7)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2826,7 +2831,7 @@
     break;
 
   case 91:
-#line 864 "/usr/homes/chet/src/bash/src/parse.y"
+#line 869 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_select_command ((yyvsp[(2) - (10)].word), REVERSE_LIST ((yyvsp[(5) - (10)].word_list), WORD_LIST *), (yyvsp[(9) - (10)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2834,7 +2839,7 @@
     break;
 
   case 92:
-#line 869 "/usr/homes/chet/src/bash/src/parse.y"
+#line 874 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_select_command ((yyvsp[(2) - (10)].word), REVERSE_LIST ((yyvsp[(5) - (10)].word_list), WORD_LIST *), (yyvsp[(9) - (10)].command), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2842,7 +2847,7 @@
     break;
 
   case 93:
-#line 876 "/usr/homes/chet/src/bash/src/parse.y"
+#line 881 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_case_command ((yyvsp[(2) - (6)].word), (PATTERN_LIST *)NULL, word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2850,7 +2855,7 @@
     break;
 
   case 94:
-#line 881 "/usr/homes/chet/src/bash/src/parse.y"
+#line 886 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_case_command ((yyvsp[(2) - (7)].word), (yyvsp[(5) - (7)].pattern), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2858,7 +2863,7 @@
     break;
 
   case 95:
-#line 886 "/usr/homes/chet/src/bash/src/parse.y"
+#line 891 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_case_command ((yyvsp[(2) - (6)].word), (yyvsp[(5) - (6)].pattern), word_lineno[word_top]);
 			  if (word_top > 0) word_top--;
@@ -2866,27 +2871,27 @@
     break;
 
   case 96:
-#line 893 "/usr/homes/chet/src/bash/src/parse.y"
+#line 898 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_function_def ((yyvsp[(1) - (5)].word), (yyvsp[(5) - (5)].command), function_dstart, function_bstart); }
     break;
 
   case 97:
-#line 896 "/usr/homes/chet/src/bash/src/parse.y"
+#line 901 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_function_def ((yyvsp[(2) - (6)].word), (yyvsp[(6) - (6)].command), function_dstart, function_bstart); }
     break;
 
   case 98:
-#line 899 "/usr/homes/chet/src/bash/src/parse.y"
+#line 904 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_function_def ((yyvsp[(2) - (4)].word), (yyvsp[(4) - (4)].command), function_dstart, function_bstart); }
     break;
 
   case 99:
-#line 903 "/usr/homes/chet/src/bash/src/parse.y"
+#line 908 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 100:
-#line 905 "/usr/homes/chet/src/bash/src/parse.y"
+#line 910 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  COMMAND *tc;
 
@@ -2918,7 +2923,7 @@
     break;
 
   case 101:
-#line 936 "/usr/homes/chet/src/bash/src/parse.y"
+#line 941 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_subshell_command ((yyvsp[(2) - (3)].command));
 			  (yyval.command)->flags |= CMD_WANT_SUBSHELL;
@@ -2926,7 +2931,7 @@
     break;
 
   case 102:
-#line 943 "/usr/homes/chet/src/bash/src/parse.y"
+#line 948 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_coproc_command ("COPROC", (yyvsp[(2) - (2)].command));
 			  (yyval.command)->flags |= CMD_WANT_SUBSHELL|CMD_COPROC_SUBSHELL;
@@ -2934,7 +2939,7 @@
     break;
 
   case 103:
-#line 948 "/usr/homes/chet/src/bash/src/parse.y"
+#line 953 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  COMMAND *tc;
 
@@ -2954,7 +2959,7 @@
     break;
 
   case 104:
-#line 965 "/usr/homes/chet/src/bash/src/parse.y"
+#line 970 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_coproc_command ((yyvsp[(2) - (3)].word)->word, (yyvsp[(3) - (3)].command));
 			  (yyval.command)->flags |= CMD_WANT_SUBSHELL|CMD_COPROC_SUBSHELL;
@@ -2962,7 +2967,7 @@
     break;
 
   case 105:
-#line 970 "/usr/homes/chet/src/bash/src/parse.y"
+#line 975 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  COMMAND *tc;
 
@@ -2982,7 +2987,7 @@
     break;
 
   case 106:
-#line 987 "/usr/homes/chet/src/bash/src/parse.y"
+#line 992 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = make_coproc_command ("COPROC", clean_simple_command ((yyvsp[(2) - (2)].command)));
 			  (yyval.command)->flags |= CMD_WANT_SUBSHELL|CMD_COPROC_SUBSHELL;
@@ -2990,117 +2995,117 @@
     break;
 
   case 107:
-#line 994 "/usr/homes/chet/src/bash/src/parse.y"
+#line 999 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_if_command ((yyvsp[(2) - (5)].command), (yyvsp[(4) - (5)].command), (COMMAND *)NULL); }
     break;
 
   case 108:
-#line 996 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1001 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_if_command ((yyvsp[(2) - (7)].command), (yyvsp[(4) - (7)].command), (yyvsp[(6) - (7)].command)); }
     break;
 
   case 109:
-#line 998 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1003 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_if_command ((yyvsp[(2) - (6)].command), (yyvsp[(4) - (6)].command), (yyvsp[(5) - (6)].command)); }
     break;
 
   case 110:
-#line 1003 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1008 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_group_command ((yyvsp[(2) - (3)].command)); }
     break;
 
   case 111:
-#line 1007 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1012 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_arith_command ((yyvsp[(1) - (1)].word_list)); }
     break;
 
   case 112:
-#line 1011 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1016 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(2) - (3)].command); }
     break;
 
   case 113:
-#line 1015 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1020 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_if_command ((yyvsp[(2) - (4)].command), (yyvsp[(4) - (4)].command), (COMMAND *)NULL); }
     break;
 
   case 114:
-#line 1017 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1022 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_if_command ((yyvsp[(2) - (6)].command), (yyvsp[(4) - (6)].command), (yyvsp[(6) - (6)].command)); }
     break;
 
   case 115:
-#line 1019 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1024 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = make_if_command ((yyvsp[(2) - (5)].command), (yyvsp[(4) - (5)].command), (yyvsp[(5) - (5)].command)); }
     break;
 
   case 117:
-#line 1024 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1029 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyvsp[(2) - (2)].pattern)->next = (yyvsp[(1) - (2)].pattern); (yyval.pattern) = (yyvsp[(2) - (2)].pattern); }
     break;
 
   case 118:
-#line 1028 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1033 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.pattern) = make_pattern_list ((yyvsp[(2) - (4)].word_list), (yyvsp[(4) - (4)].command)); }
     break;
 
   case 119:
-#line 1030 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1035 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.pattern) = make_pattern_list ((yyvsp[(2) - (4)].word_list), (COMMAND *)NULL); }
     break;
 
   case 120:
-#line 1032 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1037 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.pattern) = make_pattern_list ((yyvsp[(3) - (5)].word_list), (yyvsp[(5) - (5)].command)); }
     break;
 
   case 121:
-#line 1034 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1039 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.pattern) = make_pattern_list ((yyvsp[(3) - (5)].word_list), (COMMAND *)NULL); }
     break;
 
   case 122:
-#line 1038 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1043 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.pattern) = (yyvsp[(1) - (2)].pattern); }
     break;
 
   case 123:
-#line 1040 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1045 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyvsp[(2) - (3)].pattern)->next = (yyvsp[(1) - (3)].pattern); (yyval.pattern) = (yyvsp[(2) - (3)].pattern); }
     break;
 
   case 124:
-#line 1042 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1047 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyvsp[(1) - (2)].pattern)->flags |= CASEPAT_FALLTHROUGH; (yyval.pattern) = (yyvsp[(1) - (2)].pattern); }
     break;
 
   case 125:
-#line 1044 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1049 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyvsp[(2) - (3)].pattern)->flags |= CASEPAT_FALLTHROUGH; (yyvsp[(2) - (3)].pattern)->next = (yyvsp[(1) - (3)].pattern); (yyval.pattern) = (yyvsp[(2) - (3)].pattern); }
     break;
 
   case 126:
-#line 1046 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1051 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyvsp[(1) - (2)].pattern)->flags |= CASEPAT_TESTNEXT; (yyval.pattern) = (yyvsp[(1) - (2)].pattern); }
     break;
 
   case 127:
-#line 1048 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1053 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyvsp[(2) - (3)].pattern)->flags |= CASEPAT_TESTNEXT; (yyvsp[(2) - (3)].pattern)->next = (yyvsp[(1) - (3)].pattern); (yyval.pattern) = (yyvsp[(2) - (3)].pattern); }
     break;
 
   case 128:
-#line 1052 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1057 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.word_list) = make_word_list ((yyvsp[(1) - (1)].word), (WORD_LIST *)NULL); }
     break;
 
   case 129:
-#line 1054 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1059 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.word_list) = make_word_list ((yyvsp[(3) - (3)].word), (yyvsp[(1) - (3)].word_list)); }
     break;
 
   case 130:
-#line 1063 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1068 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = (yyvsp[(2) - (2)].command);
 			  if (need_here_doc)
@@ -3109,14 +3114,14 @@
     break;
 
   case 132:
-#line 1072 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1077 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = (yyvsp[(2) - (2)].command);
 			}
     break;
 
   case 134:
-#line 1079 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1084 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  if ((yyvsp[(1) - (3)].command)->type == cm_connection)
 			    (yyval.command) = connect_async_list ((yyvsp[(1) - (3)].command), (COMMAND *)NULL, '&');
@@ -3126,17 +3131,17 @@
     break;
 
   case 136:
-#line 1090 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1095 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = command_connect ((yyvsp[(1) - (4)].command), (yyvsp[(4) - (4)].command), AND_AND); }
     break;
 
   case 137:
-#line 1092 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1097 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = command_connect ((yyvsp[(1) - (4)].command), (yyvsp[(4) - (4)].command), OR_OR); }
     break;
 
   case 138:
-#line 1094 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1099 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  if ((yyvsp[(1) - (4)].command)->type == cm_connection)
 			    (yyval.command) = connect_async_list ((yyvsp[(1) - (4)].command), (yyvsp[(4) - (4)].command), '&');
@@ -3146,37 +3151,37 @@
     break;
 
   case 139:
-#line 1101 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1106 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = command_connect ((yyvsp[(1) - (4)].command), (yyvsp[(4) - (4)].command), ';'); }
     break;
 
   case 140:
-#line 1103 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1108 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = command_connect ((yyvsp[(1) - (4)].command), (yyvsp[(4) - (4)].command), ';'); }
     break;
 
   case 141:
-#line 1105 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1110 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 144:
-#line 1113 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1118 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.number) = '\n'; }
     break;
 
   case 145:
-#line 1115 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1120 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.number) = ';'; }
     break;
 
   case 146:
-#line 1117 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1122 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.number) = yacc_EOF; }
     break;
 
   case 149:
-#line 1131 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1136 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = (yyvsp[(1) - (1)].command);
 			  if (need_here_doc)
@@ -3192,7 +3197,7 @@
     break;
 
   case 150:
-#line 1144 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1149 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  if ((yyvsp[(1) - (2)].command)->type == cm_connection)
 			    (yyval.command) = connect_async_list ((yyvsp[(1) - (2)].command), (COMMAND *)NULL, '&');
@@ -3211,7 +3216,7 @@
     break;
 
   case 151:
-#line 1160 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1165 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  (yyval.command) = (yyvsp[(1) - (2)].command);
 			  if (need_here_doc)
@@ -3227,17 +3232,17 @@
     break;
 
   case 152:
-#line 1175 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1180 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = command_connect ((yyvsp[(1) - (4)].command), (yyvsp[(4) - (4)].command), AND_AND); }
     break;
 
   case 153:
-#line 1177 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1182 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = command_connect ((yyvsp[(1) - (4)].command), (yyvsp[(4) - (4)].command), OR_OR); }
     break;
 
   case 154:
-#line 1179 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1184 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  if ((yyvsp[(1) - (3)].command)->type == cm_connection)
 			    (yyval.command) = connect_async_list ((yyvsp[(1) - (3)].command), (yyvsp[(3) - (3)].command), '&');
@@ -3247,22 +3252,22 @@
     break;
 
   case 155:
-#line 1186 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1191 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = command_connect ((yyvsp[(1) - (3)].command), (yyvsp[(3) - (3)].command), ';'); }
     break;
 
   case 156:
-#line 1189 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1194 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 157:
-#line 1193 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1198 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 158:
-#line 1195 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1200 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  if ((yyvsp[(2) - (2)].command))
 			    (yyvsp[(2) - (2)].command)->flags ^= CMD_INVERT_RETURN;	/* toggle */
@@ -3271,7 +3276,7 @@
     break;
 
   case 159:
-#line 1201 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1206 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  if ((yyvsp[(2) - (2)].command))
 			    (yyvsp[(2) - (2)].command)->flags |= (yyvsp[(1) - (2)].number);
@@ -3280,7 +3285,7 @@
     break;
 
   case 160:
-#line 1207 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1212 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  ELEMENT x;
 
@@ -3300,7 +3305,7 @@
     break;
 
   case 161:
-#line 1224 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1229 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  ELEMENT x;
 
@@ -3321,12 +3326,12 @@
     break;
 
   case 162:
-#line 1244 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1249 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = command_connect ((yyvsp[(1) - (4)].command), (yyvsp[(4) - (4)].command), '|'); }
     break;
 
   case 163:
-#line 1246 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1251 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     {
 			  /* Make cmd1 |& cmd2 equivalent to cmd1 2>&1 | cmd2 */
 			  COMMAND *tc;
@@ -3352,28 +3357,28 @@
     break;
 
   case 164:
-#line 1269 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1274 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.command) = (yyvsp[(1) - (1)].command); }
     break;
 
   case 165:
-#line 1273 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1278 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.number) = CMD_TIME_PIPELINE; }
     break;
 
   case 166:
-#line 1275 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1280 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.number) = CMD_TIME_PIPELINE|CMD_TIME_POSIX; }
     break;
 
   case 167:
-#line 1277 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1282 "/usr/src/local/bash/bash-4.3-patched/parse.y"
     { (yyval.number) = CMD_TIME_PIPELINE|CMD_TIME_POSIX; }
     break;
 
 
 /* Line 1267 of yacc.c.  */
-#line 3377 "y.tab.c"
+#line 3382 "y.tab.c"
       default: break;
     }
   YY_SYMBOL_PRINT ("-> $$ =", yyr1[yyn], &yyval, &yyloc);
@@ -3587,7 +3592,7 @@
 }
 
 
-#line 1279 "/usr/homes/chet/src/bash/src/parse.y"
+#line 1284 "/usr/src/local/bash/bash-4.3-patched/parse.y"
 
 
 /* Initial size to allocate for tokens, and the
@@ -4736,7 +4741,7 @@
 	 not already end in an EOF character.  */
       if (shell_input_line_terminator != EOF)
 	{
-	  if (shell_input_line_size < SIZE_MAX && shell_input_line_len > shell_input_line_size - 3)
+	  if (shell_input_line_size < SIZE_MAX-3 && (shell_input_line_len+3 > shell_input_line_size))
 	    shell_input_line = (char *)xrealloc (shell_input_line,
 					1 + (shell_input_line_size += 2));
 
@@ -4845,6 +4850,16 @@
     eol_ungetc_lookahead = c;
 }
 
+char *
+parser_remaining_input ()
+{
+  if (shell_input_line == 0)
+    return 0;
+  if (shell_input_line_index < 0 || shell_input_line_index >= shell_input_line_len)
+    return '\0';	/* XXX */
+  return (shell_input_line + shell_input_line_index);
+}
+
 #ifdef INCLUDE_UNUSED
 /* Back the input pointer up by one, effectively `ungetting' a character. */
 static void
@@ -4948,13 +4963,28 @@
    which allow ESAC to be the next one read. */
 static int esacs_needed_count;
 
+static void
+push_heredoc (r)
+     REDIRECT *r;
+{
+  if (need_here_doc >= HEREDOC_MAX)
+    {
+      last_command_exit_value = EX_BADUSAGE;
+      need_here_doc = 0;
+      report_syntax_error (_("maximum here-document count exceeded"));
+      reset_parser ();
+      exit_shell (last_command_exit_value);
+    }
+  redir_stack[need_here_doc++] = r;
+}
+
 void
 gather_here_documents ()
 {
   int r;
 
   r = 0;
-  while (need_here_doc)
+  while (need_here_doc > 0)
     {
       parser_state |= PST_HEREDOC;
       make_here_document (redir_stack[r++], line_number);
@@ -5265,6 +5295,8 @@
   FREE (word_desc_to_read);
   word_desc_to_read = (WORD_DESC *)NULL;
 
+  eol_ungetc_lookahead = 0;
+
   current_token = '\n';		/* XXX */
   last_read_token = '\n';
   token_to_read = '\n';
@@ -5710,7 +5742,7 @@
          within a double-quoted ${...} construct "an even number of
          unescaped double-quotes or single-quotes, if any, shall occur." */
       /* This was changed in Austin Group Interp 221 */
-      if MBTEST(posixly_correct && shell_compatibility_level > 41 && dolbrace_state != DOLBRACE_QUOTE && (flags & P_DQUOTE) && (flags & P_DOLBRACE) && ch == '\'')
+      if MBTEST(posixly_correct && shell_compatibility_level > 41 && dolbrace_state != DOLBRACE_QUOTE && dolbrace_state != DOLBRACE_QUOTE2 && (flags & P_DQUOTE) && (flags & P_DOLBRACE) && ch == '\'')
 	continue;
 
       /* Could also check open == '`' if we want to parse grouping constructs
@@ -6317,8 +6349,8 @@
   reset_parser ();
   /* reset_parser clears shell_input_line and associated variables */
   restore_input_line_state (&ls);
-  if (interactive)
-    token_to_read = 0;
+
+  token_to_read = 0;
 
   /* Need to find how many characters parse_and_execute consumed, update
      *indp, if flags != 0, copy the portion of the string parsed into RET
@@ -8387,6 +8419,7 @@
 
   ps->expand_aliases = expand_aliases;
   ps->echo_input_at_read = echo_input_at_read;
+  ps->need_here_doc = need_here_doc;
 
   ps->token = token;
   ps->token_buffer_size = token_buffer_size;
@@ -8435,6 +8468,7 @@
 
   expand_aliases = ps->expand_aliases;
   echo_input_at_read = ps->echo_input_at_read;
+  need_here_doc = ps->need_here_doc;
 
   FREE (token);
   token = ps->token;
